README.md

```diff
@@ -113,6 +113,16 @@ python -m build --wheel # in this repository to build the wheel. Assumes you hav
 pip install dist/<generated>.whl
 ```
 
+### Vulkan support
+
+Thanks to [@thewh1teagle](https://github.com/thewh1teagle), using Vulkan is now supported:
+
+```shell
+export CMAKE_ARGS="-DGGML_VULKAN=ON"
+python -m build --wheel # in this repository to build the wheel. Assumes you have installed build with pip install build
+pip install dist/<generated>.whl
```

Comment:
Are you sure this is working ? 
You are using a flag that isn't in the Cmakelists file! 
A successful compilation does not always mean it's working! 

Did you run any tests, did you check in the output that your GPU is being used ?

Reply:
I checked it on Linux and Windows and it worked.
The GPU being used in the output.
I suggest you test it as well at least compile since the submodule didn't updated a while.

Reply:
There's another issue that I found now:
```console
Generating code
  Finished generating code
  _pywhispercpp.vcxproj -> D:\pywhispercpp\build\lib.win-amd64-cpython-39\_pywhispercpp.pyd
  Building Custom Rule D:/pywhispercpp/CMakeLists.txt
error: can't copy 'build\lib.win-amd64-cpython-39\_pywhispercpp.cp39-win_amd64.pyd': doesn't exist or not a regular file
```

Instead it generated `build\lib.win-amd64-cpython-39\_pywhispercpp.pyd`
So if I rename it to `_pywhispercpp.cp39-win_amd64.pyd` and run again it works.

Reply:
Related: https://github.com/abdeladim-s/pywhispercpp/issues/43
It also doesn't build on Windows on main branch.

Reply:
So, it worked, interesting! 
In that case can you please try to compile similar to what I did with CUDA, 
```shell
git clone --recursive https://github.com/abdeladim-s/pywhispercpp
cd pywhispercpp
GGML_VULKAN=ON pip install .
```
If it worked, then update the instructions accordingly. I wanted to update the MAC instructions as well, but I needed someone to confirm that it works.

Reply:
I will test it on my end as well, I just believe that it's safer to usually follow the official `whisper.cpp` releases, rather than compiling the bleeding edge. But it's Okey.

Reply:
> There's another issue that I found now in Windows:
> 
> ```
> Generating code
>   Finished generating code
>   _pywhispercpp.vcxproj -> D:\pywhispercpp\build\lib.win-amd64-cpython-39\_pywhispercpp.pyd
>   Building Custom Rule D:/pywhispercpp/CMakeLists.txt
> error: can't copy 'build\lib.win-amd64-cpython-39\_pywhispercpp.cp39-win_amd64.pyd': doesn't exist or not a regular file
> ```
> 
> Instead it generated `build\lib.win-amd64-cpython-39\_pywhispercpp.pyd` So if I rename it to `_pywhispercpp.cp39-win_amd64.pyd` and run again it works.
> 
> Also then it failed to import the pyd file since it's installed into `site-packages` instead of `site-packages/pywhispercpp`

Let's keep track of this issue in #43. 
Let's merge this first.

Reply:
Did you fixed it in d452d8b096b53023f4c1222dfd67367387da9d3b?

Reply:
Yes, I believe so, 
could you please give it a test on your end as well ?

Reply:
I tested now but it gave me errors

```console
git clone --recursive https://github.com/abdeladim-s/pywhispercpp
cd pywhispercpp
python -m venv venv
venv\Scripts\activate
pip install .
```

<details>

```console
 D:\pywhispercpp>pip install .
Processing d:\pywhispercpp
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting pydub
  Using cached pydub-0.25.1-py2.py3-none-any.whl (32 kB)
Collecting numpy
  Using cached numpy-2.0.2-cp39-cp39-win_amd64.whl (15.9 MB)
Collecting tqdm
  Using cached tqdm-4.66.5-py3-none-any.whl (78 kB)
Collecting requests
  Using cached requests-2.32.3-py3-none-any.whl (64 kB)
Collecting platformdirs
  Using cached platformdirs-4.2.2-py3-none-any.whl (18 kB)
Collecting colorama; platform_system == "Windows"
  Using cached colorama-0.4.6-py2.py3-none-any.whl (25 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.8-py3-none-any.whl (66 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2024.7.4-py3-none-any.whl (162 kB)
Collecting urllib3<3,>=1.21.1
  Using cached urllib3-2.2.2-py3-none-any.whl (121 kB)
Collecting charset-normalizer<4,>=2
  Using cached charset_normalizer-3.3.2-cp39-cp39-win_amd64.whl (100 kB)
Building wheels for collected packages: pywhispercpp
  Building wheel for pywhispercpp (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: 'd:\pywhispercpp\venv\scripts\python.exe' 'd:\pywhispercpp\venv\lib\site-packages\pip\_vendor\pep517\_in_process.py' build_wheel 'C:\Users\User\AppData\Local\Temp\tmpcmsot5iv'
       cwd: C:\Users\User\AppData\Local\Temp\pip-req-build-hqf07rh3
  Complete output (160 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-cpython-39
  creating build\lib.win-amd64-cpython-39\pywhispercpp
  copying .\pywhispercpp\constants.py -> build\lib.win-amd64-cpython-39\pywhispercpp
  copying .\pywhispercpp\model.py -> build\lib.win-amd64-cpython-39\pywhispercpp
  copying .\pywhispercpp\utils.py -> build\lib.win-amd64-cpython-39\pywhispercpp
  copying .\pywhispercpp\_logger.py -> build\lib.win-amd64-cpython-39\pywhispercpp
  copying .\pywhispercpp\__init__.py -> build\lib.win-amd64-cpython-39\pywhispercpp
  creating build\lib.win-amd64-cpython-39\pywhispercpp\examples
  copying .\pywhispercpp\examples\assistant.py -> build\lib.win-amd64-cpython-39\pywhispercpp\examples
  copying .\pywhispercpp\examples\livestream.py -> build\lib.win-amd64-cpython-39\pywhispercpp\examples
  copying .\pywhispercpp\examples\main.py -> build\lib.win-amd64-cpython-39\pywhispercpp\examples
  copying .\pywhispercpp\examples\recording.py -> build\lib.win-amd64-cpython-39\pywhispercpp\examples
  copying .\pywhispercpp\examples\__init__.py -> build\lib.win-amd64-cpython-39\pywhispercpp\examples
  running egg_info
  writing pywhispercpp.egg-info\PKG-INFO
  writing dependency_links to pywhispercpp.egg-info\dependency_links.txt
  writing entry points to pywhispercpp.egg-info\entry_points.txt
  writing requirements to pywhispercpp.egg-info\requires.txt
  writing top-level names to pywhispercpp.egg-info\top_level.txt
  reading manifest file 'pywhispercpp.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE'
  writing manifest file 'pywhispercpp.egg-info\SOURCES.txt'
  running build_ext
  -- Building for: Visual Studio 17 2022
  -- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22631.
  -- The C compiler identification is MSVC 19.40.33812.0
  -- The CXX compiler identification is MSVC 19.40.33812.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  CMake Deprecation Warning at pybind11/CMakeLists.txt:8 (cmake_minimum_required):
    Compatibility with CMake < 3.5 will be removed from a future version of
    CMake.

    Update the VERSION argument <min> value or use a ...<max> suffix to tell
    CMake that the project does not need compatibility with older versions.


  -- pybind11 v2.9.2
  CMake Warning (dev) at pybind11/tools/FindPythonLibsNew.cmake:98 (find_package):
    Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
    are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
    the cmake_policy command to set the policy and suppress this warning.

  Call Stack (most recent call first):
    pybind11/tools/pybind11Tools.cmake:50 (find_package)
    pybind11/tools/pybind11Common.cmake:206 (include)
    pybind11/CMakeLists.txt:200 (include)
  This warning is for project developers.  Use -Wno-dev to suppress it.

  -- Found PythonInterp: d:/pywhispercpp/venv/scripts/python.exe (found version "3.9")
  -- Found PythonLibs: C:/Users/User/AppData/Local/Programs/Python/Python39/libs/python39.lib
  -- Performing Test HAS_MSVC_GL_LTCG
  -- Performing Test HAS_MSVC_GL_LTCG - Success
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
  -- Looking for pthread_create in pthreads
  -- Looking for pthread_create in pthreads - not found
  -- Looking for pthread_create in pthread
  -- Looking for pthread_create in pthread - not found
  -- Found Threads: TRUE
  -- CMAKE_SYSTEM_PROCESSOR: AMD64
  -- x86 detected
  -- Configuring done (6.2s)
  -- Generating done (0.1s)
  CMake Warning:
    Manually-specified variables were not used by the project:

      ALLUSERSPROFILE
      APPDATA
      COMMONPROGRAMFILES
      COMMONPROGRAMFILES(X86)
      COMMONPROGRAMW6432
      COMPUTERNAME
      COMSPEC
      DRIVERDATA
      EFC_5632
      EXAMPLE_VERSION_INFO
      GOPATH
      HOMEDRIVE
      HOMEPATH
      LOCALAPPDATA
      LOGONSERVER
      NUMBER_OF_PROCESSORS
      OLLAMA_ORIGINS
      ONEDRIVE
      OS
      PATH
      PATHEXT
      PEP517_BUILD_BACKEND
      PIP_REQ_TRACKER
      PLAT
      PROCESSOR_ARCHITECTURE
      PROCESSOR_IDENTIFIER
      PROCESSOR_LEVEL
      PROCESSOR_REVISION
      PROGRAMDATA
      PROGRAMFILES
      PROGRAMFILES(X86)
      PROGRAMW6432
      PROMPT
      PSMODULEPATH
      PUBLIC
      PYTHONNOUSERSITE
      PYTHONPATH
      SESSIONNAME
      SYSTEMDRIVE
      SYSTEMROOT
      TEMP
      TMP
      USERDOMAIN
      USERDOMAIN_ROAMINGPROFILE
      USERNAME
      USERPROFILE
      VCPKG_ROOT
      VIRTUAL_ENV
      VK_SDK_PATH
      VULKAN_SDK
      WINDIR
      _OLD_VIRTUAL_PATH
      _OLD_VIRTUAL_PROMPT


  -- Build files have been written to: C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/build/temp.win-amd64-cpython-39/Release/_pywhispercpp
  MSBuild version 17.10.4+10fbfbf2e for .NET Framework

  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(541,5): warning MSB8029: The Intermediate directory or Output directory cannot reside under the Temporary directory as it could lead to issues with incremental build. [C:\Users\User\AppData\Local\Temp\pip-req-build-hqf07rh3\build\temp.win-amd64-cpython-39\Release\_pywhispercpp\ZERO_CHECK.vcxproj]
    1>Checking Build System
  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(541,5): warning MSB8029: The Intermediate directory or Output directory cannot reside under the Temporary directory as it could lead to issues with incremental build. [C:\Users\User\AppData\Local\Temp\pip-req-build-hqf07rh3\build\temp.win-amd64-cpython-39\Release\_pywhispercpp\whisper.cpp\whisper.vcxproj]
    Building Custom Rule C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/whisper.cpp/CMakeLists.txt
    ggml.c
    ggml-alloc.c
    ggml-backend.c
    ggml-quants.c
    Generating Code...
    whisper.cpp
       Creating library C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/build/temp.win-amd64-cpython-39/Release/_pywhispercpp/whisper.cpp/Release/whisper.lib and object C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/build/temp.win-amd64-cpython-39/Release/_pywhispercpp/whisper.cpp/Release/whisper.exp
    whisper.vcxproj -> C:\Users\User\AppData\Local\Temp\pip-req-build-hqf07rh3\build\temp.win-amd64-cpython-39\Release\_pywhispercpp\bin\Release\whisper.dll
  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(541,5): warning MSB8029: The Intermediate directory or Output directory cannot reside under the Temporary directory as it could lead to issues with incremental build. [C:\Users\User\AppData\Local\Temp\pip-req-build-hqf07rh3\build\temp.win-amd64-cpython-39\Release\_pywhispercpp\_pywhispercpp.vcxproj]
    Building Custom Rule C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/CMakeLists.txt
    main.cpp
       Creating library C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/build/temp.win-amd64-cpython-39/Release/_pywhispercpp/Release/_pywhispercpp.lib and object C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/build/temp.win-amd64-cpython-39/Release/_pywhispercpp/Release/_pywhispercpp.exp
    Generating code
    Finished generating code
    _pywhispercpp.vcxproj -> C:\Users\User\AppData\Local\Temp\pip-req-build-hqf07rh3\build\lib.win-amd64-cpython-39\_pywhispercpp.pyd
  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(541,5): warning MSB8029: The Intermediate directory or Output directory cannot reside under the Temporary directory as it could lead to issues with incremental build. [C:\Users\User\AppData\Local\Temp\pip-req-build-hqf07rh3\build\temp.win-amd64-cpython-39\Release\_pywhispercpp\ALL_BUILD.vcxproj]
    Building Custom Rule C:/Users/User/AppData/Local/Temp/pip-req-build-hqf07rh3/CMakeLists.txt
  error: can't copy 'build\lib.win-amd64-cpython-39\_pywhispercpp.cp39-win_amd64.pyd': doesn't exist or not a regular file
  ----------------------------------------
  ERROR: Failed building wheel for pywhispercpp
Failed to build pywhispercpp
ERROR: Could not build wheels for pywhispercpp which use PEP 517 and cannot be installed directly
WARNING: You are using pip version 20.2.3; however, version 24.2 is available.
You should consider upgrading via the 'd:\pywhispercpp\venv\scripts\python.exe -m pip install --upgrade pip' command.

```

</details

Reply:
> Can you show me the resulting files inside the build folder ?

```console
$ pywhispercpp
$ ls build
lib.win-amd64-cpython-39  temp.win-amd64-cpython-39

$ pywhispercpp
$ ls build/lib.win-amd64-cpython-39/
_pywhispercpp.pyd  pywhispercpp

$ pywhispercpp
$ ls build/temp.win-amd64-cpython-39/
Release

$ pywhispercpp
$ ls build/temp.win-amd64-cpython-39/Release/
_pywhispercpp
```



> Can you try commenting this [line](https://github.com/abdeladim-s/pywhispercpp/blob/d452d8b096b53023f4c1222dfd67367387da9d3b/setup.py#L129) and check ?

Comment out this line fix the error. But it still failed to load the extension at runtime

<details>

```console
 python .\main.py
Traceback (most recent call last):
  File "D:\pywhispercpp\pywhispercpp\examples\main.py", line 11, in <module>
    import pywhispercpp.constants as constants
  File "D:\pywhispercpp\venv\lib\site-packages\pywhispercpp\__init__.py", line 8, in <module>
    ctypes.CDLL(os.path.join(os.path.dirname(__file__), 'lib', file))
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\ctypes\__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
FileNotFoundError: Could not find module 'D:\pywhispercpp\venv\lib\site-packages\pywhispercpp\lib\_pywhispercpp.cp39-win_amd64.pyd' (or one of its dependencies). Try using the full path with constructor syntax.

find venv/ -name "*.pyd" | grep whisper
venv/Lib/site-packages/pywhispercpp/lib/_pywhispercpp.cp39-win_amd64.pyd
venv/Lib/site-packages/pywhispercpp/lib/_pywhispercpp.pyd
venv/Lib/site-packages/_pywhispercpp.pyd
```

</details>

Reply:
`_pywhispercpp.cp39-win_amd64.pyd` is the correct name basically, not sure why it has a different name. 

If you commented the line, why the file is here `venv/Lib/site-packages/pywhispercpp/lib/_pywhispercpp.cp39-win_amd64.pyd`, the lib folder should be empty!

Reply:
Maybe something is wrong with the environment!

Can you please try with Python 3.11.
Also if you can reinstall Microsft visual studio, make sure to install check the C++ package in the installation process.

Reply:
I don't think something wrong with the environemnt. llama.cpp works. whisper.cpp works.
I don't want to use Python3.11 because of stability comparing to 3.9

Reply:
This is what I used in the VM yesterday and it worked.  
I think 3.11 is stable!  There is something wrong in the build process of Python, not with actual compilation of `whisper.cpp` 

I will try to re-check the build process again!

Reply:
Here is a small workaround at least for now, you can give it a try as well. 

Inside this [for loop](https://github.com/abdeladim-s/pywhispercpp/blob/d452d8b096b53023f4c1222dfd67367387da9d3b/setup.py#L143) 

Add this
```python
if not os.path.exists(file_path):
	continue
```

Don't forget to un-comment [the line](https://github.com/abdeladim-s/pywhispercpp/blob/d452d8b096b53023f4c1222dfd67367387da9d3b/setup.py#L129)

Reply:
I installed python 3.11:

```console
py -3.11 -m venv venv
venv\Scripts\activate
python --version
Python 3.11.0
python -m pip install --upgrade pip
pip install . 
```

It build successfully then python3.11 and upgrading pip does help a little.

Did you tested `main.py` in the VM or only built it?

Reply:
Finally, the Python version was the problem then.

I ran the tests in the tests folder. No issues!

Reply:
It only build it but still failed to load. also when I execute `ldd` on the `pyd` files it says `exec format error` (on regular executable it works)

Reply:
Still failing! What's the error ? 
Could you check what's inside the lib folder ?

Reply:
![image](https://github.com/user-attachments/assets/4f13d5f8-1044-4fbd-8fd1-728c07178c07)

What's special in that project? 
Maybe we can create simple pybind11 repro to isolate the root cause

Reply:
What do you mean ? I don't get your question!

Reply:
The project https://github.com/abetlen/llama-cpp-python works and it's very similar to this one.
Maybe we can change the build system to be just like there

Reply:
No, I don't think it's similar! different approach of creating the bindings, and not just the build system. 
But the question is why? This one works as well! If it didn't work on your machine it does not mean it's not working and we need to change everything! 
Similarly, if llama-cpp-python is working for you, it does not mean it's working for everyone, check the issues there and see! 

Issues and bugs are part of the game, you are a programmer and you know that!

Reply:
Fixed in https://github.com/abdeladim-s/pywhispercpp/pull/46 and this PR works with that

Reply:
Yes, that was the idea of yesterday's patch! I located the `whisper.dll` file and tried copying it to the source directory. Just like what I did with Linux. 
I'll check #46 now.